algorithms/src/Kokkos_Random.hpp

```diff
@@ -903,50 +903,49 @@ class Random_XorShift64_Pool {
 #else
   Random_XorShift64_Pool()   = default;
 #endif
+
   Random_XorShift64_Pool(uint64_t seed) {
     num_states_ = 0;
 
     init(seed, execution_space().concurrency());
   }
 
+  Random_XorShift64_Pool(execution_space exec, uint64_t seed) {
+    num_states_ = 0;
+
+    init(exec, seed, exec.concurrency());
+  }
+
   void init(uint64_t seed, int num_states) {
     if (seed == 0) seed = uint64_t(1318319);
-    // I only want to pad on CPU like archs (less than 1000 threads). 64 is a
-    // magic number, or random number I just wanted something not too large and
-    // not too small. 64 sounded fine.
-    padding_    = num_states < 1000 ? 64 : 1;
-    num_states_ = num_states;
-
-    locks_ =
-        locks_type("Kokkos::Random_XorShift64::locks", num_states, padding_);
-    state_ = state_data_type("Kokkos::Random_XorShift64::state", num_states_,
-                             padding_);
+    init_members(num_states);
 
     typename state_data_type::HostMirror h_state =
         Kokkos::create_mirror_view(Kokkos::WithoutInitializing, state_);
     typename locks_type::HostMirror h_lock =
         Kokkos::create_mirror_view(Kokkos::WithoutInitializing, locks_);
 
-    // Execute on the HostMirror's default execution space.
-    Random_XorShift64<typename state_data_type::HostMirror::execution_space>
-        gen(seed, 0);
-    for (int i = 0; i < 17; i++) gen.rand();
-    for (int i = 0; i < num_states_; i++) {
-      int n1        = gen.rand();
-      int n2        = gen.rand();
-      int n3        = gen.rand();
-      int n4        = gen.rand();
-      h_state(i, 0) = (((static_cast<uint64_t>(n1)) & 0xffff) << 00) |
-                      (((static_cast<uint64_t>(n2)) & 0xffff) << 16) |
-                      (((static_cast<uint64_t>(n3)) & 0xffff) << 32) |
-                      (((static_cast<uint64_t>(n4)) & 0xffff) << 48);
-      h_lock(i, 0) = 0;
-    }
+    init_values_on_host(seed, num_states_, h_state, h_lock);
     deep_copy(state_, h_state);
     deep_copy(locks_, h_lock);
   }
 
-  KOKKOS_INLINE_FUNCTION Random_XorShift64<DeviceType> get_state() const {
+  void init(execution_space exec, uint64_t seed, int num_states) {
```

Comment:
Let's only have a single `init()` member function that properly passes the exec space to the view constructors with view_alloc()